docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,552 @@
+code could then look up the identifier in the module and replace the function definition as needed.
+
+See :py:class:`~tvm.relay.Module` for the definition and documentation of a module.
+
+Constant
+========
+
+This node represents a constant tensor value
+(see :py:mod:`~tvm.relay.Value` for more details).
+A constant is represented as a :py:class:`~tvm.NDArray`,
+allowing Relay to utilize TVM operators for constant evaluation.
+
+This node can also represent scalar constants, since
+scalars are tensors with a shape of :code:`()`. In the text format, numerical
+and boolean literals are thus syntactic sugar for constants encoding a
+tensor type with a rank-zero shape.
+
+See :py:class:`~tvm.relay.expr.Constant` for its definition and documentation.
+
+Tuples
+======
+
+Construction
+~~~~~~~~~~~~
+
+The tuple node builds a finite (that is, of statically known size) sequence of heterogeneous data. 
+These tuples match Python's closely. Their fixed length allows for efficient projection of their
+members.
+
+.. code-block:: python
+
+fn (%a : Tensor[(10, 10), float32], %b : float32, %c : Tensor[(100, 100), float32]) {
+    let %tup = (%a, %b);     // type: (Tensor[(10, 10), float32], float32)
+    ((%tup.0 + %tup.1), %c)  // type: (Tensor[(10, 10), float32], Tensor[(100, 100), float32])
+}
+
+See :py:class:`~tvm.relay.expr.Tuple` for its definition and documentation.
+
+Projection
+~~~~~~~~~~
+
+A tuple must be indexed by an integer constant in order to extract a
+particular member of the tuple. Projections are 0-indexed.
+
+For example, the below projection evaluates to :code:`%b`:
+
+.. code-block:: python
+
+   (%a, %b, %c).1
+
+See :py:class:`~tvm.relay.expr.TupleGetItem` for its definition and documentation.
+
+Let Bindings
+============
+
+A :code:`let` binding is an immutable local variable binding,
+allowing the user to bind an expression to a name.
+
+A :code:`let` binding contains a local variable,
+an optional type annotation, a value, and a body expression
+that may reference the bound identifier. If a type annotation
+on the bound variable is omitted, Relay attempts to infer the
+most general type permitted for the variable.
+
+The bound variable in a :code:let expression is only in scope 
+in its body, except when the variable defines a function expression.
+When a :code:let expression creates a function, the variable is also
+in scope in its value to allow for recursively defined functions 
+(see the previous subsection).
+
+The value of a :code:`let` binding is the value of the final expression
+after evaluating the bindings it depends on. For example, in the
+following example the entire expression evaluates to a tensor
+of shape (10, 10) where all elements are 2:
+
+.. code-block:: python
+
+   let %x : Tensor[(10, 10), float32] = Constant(1, (10, 10), float32);
+   %x + %x
+
+A sequence of :code:`let` bindings can be considered as a dataflow graph,
+where the bindings are a series of sub-graphs connected
```

Comment:
this could use a picture, but maybe we should punt on that

Reply:
Pictures may be useful for publications, but we could always add them later.